### ISSUE TITLE
Refactor enter-chroot into separate modules

### ISSUE THREAD

**Issue by ccaapton** [dnschneid/crouton#647]
Currently the enter-chroot serves as the monolithic script for many purposes:
1. Parsing all the arguments and env variables
2. Preparing the host, including chgrp video "$fbc", hung_task_panic, drm_relax_file
3. Run setup script if it exists and yet we're logging in. I don't think this task should be implement in enter-chroot. Maybe something called post_bootstrap is more appropriate. 
4. Preparing the guest before chroot. Including making a lot of softlinks inside chroot, copy files inside chroot
   A lot of the copying/softlinks can be replaced by bindmount(for regular files), personally I think it is a cleaner solution.
5. Bindmount, tempfs mount
6. Run some stuff using chrootcmd, like groupadd/groupmod. Again, we don't need this if we just need a one-shot task. We should also delay the task untill the last mount, just before the intended task,  to avoid too many chroot/unchroot syscalls.
7. Formally enters the chroot, run rc.local if neccessary, and then the main task.

The current script makes it difficult to adapt to alternative chroot solutions, like minijail0, or extend to separate namespaces. A better modular division of these functionalities will help to do better experiments with namespaces, lxc etc. It can also reduce maintenance cost.

Here are some of my suggestions:

a. all the bind mounts and chroot should be atomic. This is the behavior of minijail0, it can also stop us from doing stupid things in the middle.

b. Lazy bind mount. There are many bindmount() inside if conditions, which makes a) hard. However we can solve this by only recording the task inside bindmount(){}, like 

```
#!/usr/bin/env python
def bindmount(src,target,opt):
    _lazy_binds+=[src,target,opt]
```

c. Avoid copying/ files into chroot, as well as softlinking things. Using bind mount instead.

d. Stop binding /tmp, as most init program will purge this directory. Bind it somewhere else as readonly is preferred. 

e. The lazy task idea also works for chrootcmd.

f. Tmpfs should be setup after bind/chroot, maybe in fstab, or rc.local

g. Again, I appeal for python. Most of the job in the current script is checking env varialbes, checking file existences, and string manipulations. Python can complete them easily, and is more readable/writable. ChromeOS does include python at /usr/bin/python, at least in our dev-mode. We only care about things in dev-mode, right?

To summarize, we can divide the core functionality of enter-chroot into 4 parts:
1. host_prepare
2. lazy collect bind information and chrootcmds, to be used later
3. bindmount and chroot. This part should be extensible so minijail/lxc can be used.
4. post-chroot in guest. This part should be done in rc.local, fstab, or other init scripts.
